libp2p/protocols/pubsub/pubsubpeer.nim

```diff
@@ -181,6 +182,24 @@ proc handle*(p: PubSubPeer, conn: Connection) {.async.} =
     debug "exiting pubsub read loop",
       conn, peer = p, closed = conn.closed
 
+proc disconnect(p: PubSubPeer) {.async.} =
```

Comment:
this doesn't disconnect the peer - it just closes the send connection

Reply:
I can't disconnect the peer from `PubSubPeer`, it's available only in GossipSub and I don't have a reference to it here cause of circular dependency limitation.

Reply:
It seems we emit a `Disconnected` event that doesn't mean a disconnection.

Reply:
Let me know if I goit it right.

1. We reset the stream and close the underlying connection here, but don't raise any error so at the time no error happens https://github.com/vacp2p/nim-libp2p/blob/28609597d104a9be880ed5e1648e1ce18ca9dc38/libp2p/muxers/mplex/lpchannel.nim#L220
2. Next time we try to write to the stream this error happens https://github.com/vacp2p/nim-libp2p/blob/e0bcc5fed20ea1b6c8eea605a4ad4caa275b3fb3/libp2p/protocols/pubsub/pubsubpeer.nim#L181
3. Then this is execute https://github.com/vacp2p/nim-libp2p/blob/e0bcc5fed20ea1b6c8eea605a4ad4caa275b3fb3/libp2p/protocols/pubsub/pubsubpeer.nim#L229

So the only difference now is that the underlying connection hasn't been closed?

Reply:
> So the only difference now is that the underlying connection hasn't been closed?

yes - closing the underlying ensures that all ongoing concurrent reads, writes etc are aborted and we get rid of the peer - this is important because we don't know what state gossipsub is in if we just abort here - they might have received control messages or not so closing the connection entirely is one of the safest and least bad things you can do.

Reply:
there exists an ugly-hack solution: drain the non-prio queue and send all non-prio messages as priority messages - this will _likely_ cause mplex to disconnect

Reply:
But how can I close the underlying connection here? The conn here is actually a stream (btw what about calling it a stream when it is one?) and `Connection` doesn't expose the underlying transport connection.

Reply:
@dryajov knows more about the stream vs connection naming, maybe

Reply:
I could rename `PubSubPeerEventKind.Disconnected` to `PubSubPeerEventKind.StreamClosed` and create a `PubSubPeerEventKind.Disconnect`. Then `GossipSub` would disconnect the peer when receiving the latter. This would be the closest to the current behavior, I guess.  Another option would be to disconnect the peer here https://github.com/vacp2p/nim-libp2p/blob/1734160f8a3e759ea03145878b3e7f19f2addbad/libp2p/protocols/pubsub/gossipsub.nim#L185, but it would change the current behavior.

Reply:
> But how can I close the underlying connection here? 

see suggested send-all-non-prio-messages trick - it should work because if there's at least 1 in-flight future, it will cause mplex to do the disconnection.

Reply:
Wouldn't emit an event that tells `GossipSub` to disconnect the peer be simpler and easier to understand?

Reply:
but I think just writing the whole queue might not trigger the disconnection if the futures complete fast enough.